File: app/src/main/java/com/weathercock/cleanarchitecture/data/dao/WeatherDao.java
package com.weathercock.cleanarchitecture.data.dao;

import android.support.annotation.NonNull;

import com.weathercock.cleanarchitecture.data.entity.ForecastEntity;
import com.weathercock.cleanarchitecture.data.entity.ImageEntity;
import com.weathercock.cleanarchitecture.data.entity.TemperatureUnitEntity;
import com.weathercock.cleanarchitecture.data.entity.WeatherEntity;

import io.realm.Realm;

/**
 * Weather Dao
 * <p>
 * Created by tamura_k on 2016/06/03.
 */
public class WeatherDao extends AbsDao {

    public WeatherDao(@NonNull Realm realm) {
        super(realm);
    }

    public WeatherEntity find(int cityId) {
        return getRealm().where(WeatherEntity.class)
                .equalTo("cityId", cityId)
                .findFirst();
    }

    public boolean exist(int cityId) {
        return (0 <= getRealm().where(WeatherEntity.class)
                .equalTo("cityId", cityId)
                .count());
    }

    public void insert(WeatherEntity weather) {
        getRealm().copyToRealm(weather);
    }

    public void delete(int cityId) {
        WeatherEntity weather = find(cityId);
        if (weather != null) {
            for (ForecastEntity forecast : weather.getForecasts()) {
                TemperatureUnitEntity max = forecast.getTemperature().getMax();
                if (max != null) max.deleteFromRealm();
                TemperatureUnitEntity min = forecast.getTemperature().getMin();
                if (min != null) min.deleteFromRealm();
                forecast.getTemperature().deleteFromRealm();
                forecast.getImage().deleteFromRealm();
            }
            weather.getForecasts().deleteAllFromRealm();
            weather.getCopyright().getProvider().deleteAllFromRealm();
            ImageEntity image = weather.getCopyright().getImage();
            if (image != null) image.deleteFromRealm();
            weather.getCopyright().deleteFromRealm();
            weather.getLocation().deleteFromRealm();
            weather.getDescription().deleteFromRealm();
            weather.getPinpointLocations().deleteAllFromRealm();
            weather.deleteFromRealm();
        }
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/data/database/dao/weather/WeatherDao.kt
package com.kazakago.cleanarchitecture.data.database.dao.weather

import android.arch.persistence.room.Dao
import android.arch.persistence.room.Delete
import android.arch.persistence.room.Insert
import android.arch.persistence.room.Query
import com.kazakago.cleanarchitecture.data.database.entity.weather.DescriptionEntity
import com.kazakago.cleanarchitecture.data.database.entity.weather.ForecastEntity
import com.kazakago.cleanarchitecture.data.database.entity.weather.LocationEntity
import com.kazakago.cleanarchitecture.data.database.entity.weather.WeatherEntity

@Dao
interface WeatherDao {

    @Query("SELECT * FROM Weather WHERE city_id == (:cityId) LIMIT 1")
    fun findWeather(cityId: String): WeatherEntity?

    @Insert
    fun insert(weather: WeatherEntity)

    @Delete
    fun delete(weather: WeatherEntity)

    @Query("SELECT * FROM Location WHERE city_id == (:cityId) LIMIT 1")
    fun findLocation(cityId: String): LocationEntity?

    @Insert
    fun insert(location: LocationEntity)

    @Delete
    fun delete(location: LocationEntity)

    @Query("SELECT * FROM Description WHERE city_id == (:cityId) LIMIT 1")
    fun findDescription(cityId: String): DescriptionEntity?

    @Insert
    fun insert(description: DescriptionEntity)

    @Delete
    fun delete(description: DescriptionEntity)

    @Query("SELECT * FROM Forecast WHERE city_id == (:cityId)")
    fun findForecasts(cityId: String): List<ForecastEntity>?

    @Insert
    fun insert(forecasts: List<ForecastEntity>)

    @Delete
    fun delete(forecasts: List<ForecastEntity>)

}
